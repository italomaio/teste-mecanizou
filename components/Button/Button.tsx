import React, { forwardRef } from "react";
import clsx from "clsx";

type ButtonVariant = "primary" | "secondary" | "danger" | "ghost" | "outline";
type ButtonSize = "sm" | "md" | "lg" | "icon";

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: ButtonVariant;
  size?: ButtonSize;
  icon?: React.ReactNode;
}

const variantClasses: Record<ButtonVariant, string> = {
  primary:
    "bg-sapphire text-white hover:bg-sapphire-600 border border-sapphire-600",
  secondary: "bg-hot-cinnamon text-white hover:bg-sapphire",
  outline: "",
  danger: "bg-red-500 text-white hover:bg-red-600 border border-red-600",
  ghost: "hover:bg-accent hover:text-accent-foreground",
};

const sizeClasses: Record<ButtonSize, string> = {
  sm: "px-2 py-1 text-sm",
  md: "px-4 py-2 text-base",
  lg: "px-6 py-3 text-lg",
  icon: "p-2 text-base",
};

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    { variant = "primary", size = "md", icon, className, children, ...props },
    ref
  ) => {
    return (
      <button
        ref={ref}
        className={clsx(
          "rounded focus:outline-none inline-flex items-center justify-center gap-2 transition-colors duration-300 cursor-pointer font-medium text-md",
          variantClasses[variant],
          sizeClasses[size],
          className
        )}
        {...props}
      >
        {icon && <span className="flex items-center">{icon}</span>}
        {children}
      </button>
    );
  }
);

Button.displayName = "Button";

export default Button;
