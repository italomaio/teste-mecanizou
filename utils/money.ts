export const formatter = new Intl.NumberFormat("pt-BR", {
  style: "currency",
  currency: "BRL",
  minimumFractionDigits: 2,
});

export function calculateDiscountPrice(
  price: number,
  discountPercentage: number
) {
  const discountFactor = discountPercentage / 100;
  const newPrice = price * (1 - discountFactor);

  return {
    oldPrice: formatter.format(price),
    newPrice: formatter.format(newPrice),
  };
}
