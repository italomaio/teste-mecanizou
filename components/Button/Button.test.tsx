import { render } from "@/tests/utils";
import { screen, fireEvent } from "@testing-library/react";
import { Search } from "lucide-react";

import Button from "./Button";

describe("Button Component", () => {
  it("Should render button with default props", () => {
    render(<Button>Click Me</Button>);
    const button = screen.getByRole("button", { name: /click me/i });
    expect(button).toBeInTheDocument();
    expect(button).toHaveClass("text-white bg-sapphire");
  });

  it("Should apply the correct variant class", () => {
    render(<Button variant="secondary">Login</Button>);
    const button = screen.getByRole("button", { name: /Login/i });
    expect(button).toHaveClass("bg-hot-cinnamon text-white");
  });

  it("Should apply the correct size class", () => {
    render(<Button size="lg">Large Button</Button>);

    const button = screen.getByRole("button", { name: /large button/i });
    expect(button).toHaveClass("px-6 py-3 text-lg");
  });

  it("Should render icon button", () => {
    render(
      <Button size="icon">
        <Search />
      </Button>
    );

    const button = screen.getByRole("button");
    expect(button.querySelector("svg")).toBeInTheDocument();
    expect(button).toHaveClass("p-2 text-base");
  });

  it("Should handle click events", () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click Me</Button>);
    const button = screen.getByRole("button", { name: /click me/i });

    fireEvent.click(button);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it("Should apply custom className", () => {
    render(<Button className="custom-class">Custom</Button>);
    const button = screen.getByRole("button", { name: /custom/i });
    expect(button).toHaveClass("custom-class");
  });

  it("Should render as disabled when disabled prop is passed", () => {
    render(<Button disabled>Disabled</Button>);
    const button = screen.getByRole("button", { name: /disabled/i });
    expect(button).toBeDisabled();
  });

  it("Should match snapshot", () => {
    const { asFragment } = render(<Button>Snapshot Test</Button>);
    expect(asFragment()).toMatchSnapshot();
  });
});
