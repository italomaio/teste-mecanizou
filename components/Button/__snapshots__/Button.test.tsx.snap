// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button Component Should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="rounded focus:outline-none inline-flex items-center justify-center gap-2 transition-colors duration-300 cursor-pointer font-medium text-md bg-sapphire text-white hover:bg-sapphire-600 border border-sapphire-600 px-4 py-2 text-base"
  >
    Snapshot Test
  </button>
</DocumentFragment>
`;
